flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarSubmissionITCase.java

```diff
@@ -0,0 +1,223 @@
+import org.apache.flink.runtime.rest.messages.JobPlanInfo;
+import org.apache.flink.runtime.testingUtils.TestingUtils;
+import org.apache.flink.runtime.webmonitor.RestfulGateway;
+import org.apache.flink.runtime.webmonitor.TestingDispatcherGateway;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+import org.apache.flink.util.ExceptionUtils;
+
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.net.ConnectException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+import static org.hamcrest.Matchers.containsString;
+
+/**
+ * Tests the entire lifecycle of a jar submission.
+ */
+public class JarSubmissionITCase {
+
+	@Rule
+	public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+	@Test
+	public void testJarSubmission() throws Exception {
+		final TestingDispatcherGateway restfulGateway = new TestingDispatcherGateway.Builder().build();
+		final JarHandlers handlers = new JarHandlers(temporaryFolder.newFolder().toPath(), restfulGateway);
+		final JarUploadHandler uploadHandler = handlers.uploadHandler;
+		final JarListHandler listHandler = handlers.listHandler;
+		final JarPlanHandler planHandler = handlers.planHandler;
+		final JarRunHandler runHandler = handlers.runHandler;
+		final JarDeleteHandler deleteHandler = handlers.deleteHandler;
+
+		// targetDir property is set via surefire configuration
+		final Path originalJar = Paths.get(System.getProperty("targetDir")).resolve("test-program.jar");
+		final Path jar = Files.copy(originalJar, temporaryFolder.getRoot().toPath().resolve("test-program.jar"));
+
+		final String storedJarPath = uploadJar(uploadHandler, jar, restfulGateway);
+		final String storedJarName = Paths.get(storedJarPath).getFileName().toString();
+
+		final JarListInfo postUploadListResponse = listJars(listHandler, restfulGateway);
+		Assert.assertEquals(1, postUploadListResponse.jarFileList.size());
+		final JarListInfo.JarFileInfo listEntry = postUploadListResponse.jarFileList.iterator().next();
+		Assert.assertEquals(jar.getFileName().toString(), listEntry.name);
+		Assert.assertEquals(storedJarName, listEntry.id);
+
+		final JobPlanInfo planResponse = showPlan(planHandler, storedJarName, restfulGateway);
+		// we're only interested in the core functionality so checking for a small detail is sufficient
+		Assert.assertThat(planResponse.getJsonPlan(), containsString("TestProgram.java:29"));
+
+		try {
+			runJar(runHandler, storedJarName, restfulGateway);
+			Assert.fail("We assume the actual job submission to fail.");
+		} catch (Exception e) {
+			final Optional<ConnectException> expected = ExceptionUtils.findThrowable(e, ConnectException.class);
```

Comment:
we can make this part a lot nicer by starting a `BlobServer´ like we do in the `JobSubmitHandlerTest` and setting a `submitFunction` on the `TestingDispatcherGateway`.

Will add this later.